include/crow/middlewares/cors.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+#include "crow/http_request.h"
+#include "crow/http_response.h"
+
+namespace crow
+{
+    struct CORSHandler;
+
+    // CORSRules is used for tuning cors policies
+    struct CORSRules
+    {
+        friend struct crow::CORSHandler;
+
+        // Set Access-Control-Allow-Origin. Default is "*"
+        CORSRules& origin(const std::string& origin)
+        {
+            origin_ = origin;
+            return *this;
+        }
+
+        // Set Access-Control-Allow-Methods. Default is "*"
+        CORSRules& methods(crow::HTTPMethod method)
+        {
+            add_list_item(methods_, crow::method_name(method));
+            return *this;
+        }
+
+        // Set Access-Control-Allow-Methods. Default is "*"
+        template<typename... Methods>
+        CORSRules& methods(crow::HTTPMethod method, Methods... method_list)
+        {
+            add_list_item(methods_, crow::method_name(method));
+            methods(method_list...);
+            return *this;
+        }
+
+        // Set Access-Control-Allow-Headers. Default is "*"
+        CORSRules& headers(const std::string& header)
+        {
+            add_list_item(headers_, header);
+            return *this;
+        }
+
+        // Set Access-Control-Allow-Headers. Default is "*"
+        template<typename... Headers>
+        CORSRules& headers(const std::string& header, Headers... header_list)
+        {
+            add_list_item(headers_, header);
+            headers(header_list...);
+            return *this;
+        }
+
+        // Set Access-Control-Max-Age. Default is none
+        CORSRules& max_age(int max_age)
+        {
+            max_age_ = std::to_string(max_age);
+            return *this;
+        }
+
+        // Enable Access-Control-Allow-Credentials
+        CORSRules& allow_credentials()
+        {
+            allow_credentials_ = true;
+            return *this;
+        }
+
+        // Ignore CORS and don't send any headers
+        void ignore()
+        {
+            ignore_ = true;
+        }
+
+    private:
+        // build comma separated list
+        void add_list_item(std::string& list, const std::string& val)
+        {
+            if (list == "*") list = "";
+            if (list.size() > 0) list += ", ";
+            list += val;
+        }
+
+        // Set header `key` to `value` if it is not set
+        void set_header(const std::string& key, const std::string& value, crow::response& res)
+        {
+            if (value.size() == 0) return;
+            if (!get_header_value(res.headers, key).empty()) return;
```

Comment:
Headers are **not** overwritten, this means that a handler can change the CORS behaviour if it wants to